lib/solarwinds_apm/support/resource_detector.rb

```diff
@@ -0,0 +1,129 @@
+    UAMS_CLIENT_URL = 'http://127.0.0.1:2113/info/uamsclient'
+    UAMS_CLIENT_ID_FIELD = 'uamsclient_id'
+
+    def self.detect
+      attributes = ::OpenTelemetry::SDK::Resources::Resource.create({})
+      attributes = attributes.merge(detect_uams_client_id)
+      attributes = attributes.merge(detect_k8s_atttributes)
+      attributes.merge(from_upstream_detector)
+    end
+
+    def self.detect_uams_client_id
+      uams_client_id = nil
+      if File.exist?(UAMS_CLIENT_PATH)
+        uams_client_id = File.read(UAMS_CLIENT_PATH).strip
+      else
+        url = URI(UAMS_CLIENT_URL)
+
+        response = nil
+        ::OpenTelemetry::Common::Utilities.untraced do
+          http = Net::HTTP.new(url.host, url.port)
+          request = Net::HTTP::Get.new(url)
+          response = http.request(request)
+        end
+
+        if response&.code.to_i == 200
+          uams_metadata = JSON.parse(response.body)
+          uams_client_id = uams_metadata&.fetch(UAMS_CLIENT_ID_FIELD)
+        end
+      end
+
+      resource_attributes = {
+        'sw.uams.client.id' => uams_client_id
+      }
+
+      ::OpenTelemetry::SDK::Resources::Resource.create(resource_attributes)
+    rescue StandardError => e
+      SolarWindsAPM.logger.debug "#{self.class}/#{__method__}] detect_uams_client_id failed. Error: #{e.message}."
+      ::OpenTelemetry::SDK::Resources::Resource.create({})
+    end
+
+    def self.detect_k8s_atttributes
+      return ::OpenTelemetry::SDK::Resources::Resource.create({}) unless ENV['KUBERNETES_SERVICE_HOST'] && ENV['KUBERNETES_SERVICE_PORT']
+
+      resource_attributes = {}
+      pod_name = File.read(K8S_PODNAME_PATH).strip if File.exist?(K8S_PODNAME_PATH)
+      namespace = File.read(K8S_NAMESPACE_PATH).strip if File.exist?(K8S_NAMESPACE_PATH)
+      pod_uid = nil
+      token = nil
+
+      token = File.read(K8S_TOKEN_PATH).strip if File.exist?(K8S_TOKEN_PATH)
+
+      if token && pod_name && namespace
+        url = URI("https://kubernetes.default.svc/api/v1/namespaces/#{namespace}/pods/#{pod_name}")
+
+        response = nil
+        ::OpenTelemetry::Common::Utilities.untraced do
+          http = Net::HTTP.new(url.host, url.port)
+          http.use_ssl = true
+          http.verify_mode = OpenSSL::SSL::VERIFY_PEER
+
+          request = Net::HTTP::Get.new(url)
+          request['Authorization'] = "Bearer #{token}"
+          request['Accept'] = 'application/json'
+
+          response = http.request(request)
+        end
+
+        if response&.code.to_i == 200
+          pod_metadata = JSON.parse(response.body)
+          pod_uid = pod_metadata&.fetch('metadata')&.fetch('uid')
+        end
+      end
+
+      resource_attributes['k8s.namespace.name'] = namespace
+      resource_attributes['k8s.pod.name'] = pod_name
+      resource_attributes['k8s.pod.uid'] = pod_uid
+
+      resource_attributes.compact!
+      ::OpenTelemetry::SDK::Resources::Resource.create(resource_attributes)
+    end
```

Comment:
It seems upstream does not yet detect the [service instance](https://github.com/solarwinds-cloud/trace/blob/resource-attributes/docs/specs/resource-attributes.md#service-instance) resource attribute, so we'll need to implement our own here too.

Reply:
is the `service.instance.id` same as `k8s.pod.uid`?

Reply:
No, `service.instance.id` should be a globally unique [UUID](https://opentelemetry.io/docs/specs/semconv/resource/#service), while `k8s.pod.uid` only identifies the pod, not the containers (thus service instances) in the pod.

Reply:
`service.instance.id` should be generated by language sdk? I will implement one since current otel ruby sdk doesn't generate such value.

Reply:
Yes, you can see the JS [implementation here](https://github.com/open-telemetry/opentelemetry-js/blob/3c452f4b20f69383167712553a8c869d36b388dc/packages/opentelemetry-resources/src/detectors/platform/node/ServiceInstanceIdDetector.ts#L17). Might be nice to contribute to the upstream [default](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/sdk/lib/opentelemetry/sdk/resources/resource.rb) detector.

Reply:
Updated. Ruby seems have `SecureRandom.uuid` that can generate uuid